active/0000-structures.md

```diff
@@ -0,0 +1,150 @@
+  },
+  tokens: {
+    // values used in creating tokens
+    name: function(options) {
+      return options.blueprintName;
+    },
+    path: function(options) {
+      return path.join(options.podPath, options.dasherizedModuleName);
+    },
+    test: function(options) {
+      return options.blueprintName;
+    }
+  },
+  resolver: {
+    // rules for resolving go here, still need to figure out how this would look.
+  }
+};
+
+```
+Project would get a couple convenience methods for looking up and adding structures.
+
+```js
+Project.prototype.lookupStructures = function() {
+  // pseudo lookup code to get folder contents and return as an array
+  var structures = glob.sync(path.join(this.root,'structures','*'));
+
+  return structures;
+}
+
+Project.prototype.buildStructure = function(path) {
+  var structureOptions = require(path);
+  var Constructor = Structure.extend(structureOptions);
+  return new Constructor;
+}
+
+Project.prototype.structures = function(options){
+  if (this._structures) { return this._structures; }
+  // lookup structures in project
+  this._structures = this.lookupStructures()
+    .map(this.buildStructure).bind(this);
+
+  return this._structures;
+}
+```
+Blueprint changes would require a way to get the available structures to determine the current structure. From there the currentStructure is used to extract the values for fileMapTokens.
+
+```js
+Blueprint.prototype.getStructure = function(options) {
+  var availableStructures = this.project.structures();
+  // if the structure passed in options isn't supported fall back to default
+  if (availableStructures[options.structure].isSupported(this)) {
+    return availableStructures[options.structure];
+  }
+  return availableStructures['basic'];
+}
+
+Blueprint.prototype._fileMapTokens = function(options) {
+  var currentStructure = this.getStructure(options);
+  var standardTokens = {
+    __name__: function(options) {
+      return currentStructure.getToken('name',options);
+    },
+    __path__: function(options) {
+      return currentStructure.getToken('path',options);
+    },
+    __test__: function(options) {
+      return currentStructure.getToken('test',options);
+    }
+  };
+
+  var customTokens = this.fileMapTokens(options) || options.fileMapTokens || {};
+  return merge(standardTokens, customTokens);
+};
+
+```
+
+The resolver would require some modification to allow the ability to pickup new rules through ember-cli, and I haven't completely investigated that part yet.
+
+```js
+// broccoli/ember-app.js
+
+// something to insert the resolver rules into the app resolver goes here...
+
+```
+
+
+### Steps to implement
+After the Structure class has been completed, the first step of this would probably be to extract pod lookup patterns from the resolver into a structure and implement them in ember-cli
+
+
+# Drawbacks
+
+It may be a overly-complicated solution for a problem that we don't have many use cases for yet.
```

Comment:
Yes I agree, although it may be a nice internal refactoring. We should only proceed if this ends up feeling like a net benefit. I believe this is absolutely possible though.